repair/row_level.cc

```diff
@@ -979,7 +998,15 @@ class repair_meta {
                 _remote_sharder,
                 _master_node_shard_config.shard,
                 _seed,
-                repair_reader::is_local_reader(_repair_master || _same_sharding_config));
+                std::invoke([this]() {
+                    if (_repair_master || _same_sharding_config) {
+                        return repair_reader::read_strategy::local;
+                    }
+                    if (_estimated_partitions <= 100) {
```

Comment:
Instead of this hard-coded limit of 100, it would be better to compare against something like the number of ranges. Note that discarding a partition also incurs a possible `smp::submit_to()` to each shards, because the filter calls `next_partition()`. So in essence, this optimization relies on getting away with fewer `next_partition()` then `fast_forward_to()`.

Reply:
The row level repair always works on a single token range. The sharder will convert the single range into lots of subranges. The number of token ranges depends on the token range, shard number and msb. Assume a single token range is translated to N subranges.

With current split approach:
- Each sub reader of the multishard reader will work on N sub ranges. 
- No need to drop any partitions
 
With the filter approach:
- Each sub reader of the multishard reader will work on 1 original ranges. 
- We need to drop any partitions that does not needed

It is possible that the subrange we need contains few data but the unwanted subranges contains lots of data. It is a big waste to read those partitions. On the other hand, the with first approach, we worked on more sub ranges, but at least we do not waste disk io to read unwanted partitions and drop.  

It is hard to say which one is more efficient and select which one to use is better.

Reply:
@gusev-p Do you see noticeable difference with release / dev build with your patch?

Reply:
@asias note that `fast_forward_to()` (used to skip between ranges) is not free either. It involves I/O (possibly) to read the index. Filtering also doesn't mean the entire partition is read (which could be a huge waste indeed). The reader will read a buffer worth of data from a partition, or if the partition is small, multiple partitions until the buffer is full. The filter will then instruct the reader to skip to the next partition. If the partition is small, and fits into the buffer entirely, the reader just has to discard data from the buffer. If the partition is large, only a buffer-full is thrown away, and the reader will do a skip using the index (possibly doing I/O).

So, neither variant is free and there can be cases where either will be cheaper than the other. The logic proposed in this patch *does* have the possibility to speed up repair. What I'm not confident is the heuristics it bases this choice upon.

Reply:
> @gusev-p Do you see noticeable difference with release / dev build with your patch?

```
with optimisation:
    16:52:21.951 INFO> [release/topology_custom.test_topology_smp.1] starting server at host 127.75.65.3 in scylla-2...
    16:52:24.608 INFO> [release/topology_custom.test_topology_smp.1] started server at host 127.75.65.3 in scylla-2, pid 1154689

    24608 - 21951 = 2657

without optimisation:
    17:01:25.408 INFO> [release/topology_custom.test_topology_smp.1] starting server at host 127.77.171.3 in scylla-2...
    17:01:28.987 INFO> [release/topology_custom.test_topology_smp.1] started server at host 127.77.171.3 in scylla-2, pid 1155934

    28987 - 25408 = 3579
```

This optimisation is mostly for debug builds.